### PR TITLE
Security & reliability hardening for 4.4.0 (Tier 1 remediation, MFA #313, session regen)

### LINKED ISSUE #313: MFA: second-factor login replaces session authentication instead of merging factor authorities (flow can never complete)
## Summary

With `user.mfa.enabled=true` and `factors: PASSWORD, WEBAUTHN`, completing the WebAuthn challenge after a password login **replaces** the session's authentication instead of **adding** the `WEBAUTHN` factor to it. The session goes from `satisfiedFactors=[PASSWORD]` to `satisfiedFactors=[WEBAUTHN]` — never both — so `AllRequiredFactorsAuthorizationManager` keeps denying access and the user bounces between the two factor entry points forever. The MFA flow introduced in #272 can never complete.

## Reproduction

Found while testing the companion demo app PR ([SpringUserFrameworkDemoApp#60](https://github.com/devondragon/SpringUserFrameworkDemoApp/pull/60)) with a Playwright E2E test using Chromium's CDP virtual authenticator:

1. Enable MFA (`user.mfa.enabled=true`, `factors: PASSWORD, WEBAUTHN`)
2. Log in with username/password → `GET /user/mfa/status` reports `satisfiedFactors=["PASSWORD"], missingFactors=["WEBAUTHN"]`
3. Complete a WebAuthn assertion (`POST /login/webauthn`)
4. `GET /user/mfa/status` now reports `satisfiedFactors=["WEBAUTHN"], missingFactors=["PASSWORD"]`, `fullyAuthenticated=false`
5. Any protected request is denied for the missing `PASSWORD` factor and redirected to `passwordEntryPointUri`; logging in with the password again drops `WEBAUTHN`, and so on

## Root cause

Spring Security 7's factor accumulation is implemented in `AbstractAuthenticationProcessingFilter.doFilter`: when its `mfaEnabled` flag is set and the new authentication has the same name as the current one, the new authentication is rebuilt via `toBuilder()` with the **union** of both authorities (this is how `FactorGrantedAuthority`s accumulate across logins).

That flag is off by default. It is normally turned on by `@EnableMultiFactorAuthentication`, whose import (`EnableMfaFiltersConfiguration`) registers a `BeanPostProcessor` that calls `setMfaEnabled(true)` on every `AbstractAuthenticationProcessingFilter`, `AuthenticationFilter`, `AbstractPreAuthenticatedProcessingFilter`, and `BasicAuthenticationFilter`.

The framework's `MfaConfiguration` replicates the *authorization* half of that annotation (it builds the `DefaultAuthorizationManagerFactory` with `AllRequiredFactorsAuthorizationManager` from `user.mfa.factors`), but nothing performs the *filter* half — `mfaEnabled` is never set, so authorities are never merged and each login replaces the previous factor.

## Suggested fix

When `user.mfa.enabled=true`, `MfaConfiguration` should also register the filter post-processor, e.g.:

```java
@Bean
@ConditionalOnProperty(name = "user.mfa.enabled", havingValue = "true")
static BeanPostProcessor mfaFilterEnabler() {
    return new BeanPostProcessor() {
        @Override
        public Object postProcessAfterInitialization(Object bean, String beanName) {
            if (bean instanceof AbstractAuthenticationProcessingFilter filter) {
                filter.setMfaEnabled(true);
            }
            if (bean instanceof AuthenticationFilter filter) {
                filter.setMfaEnabled(true);
            }
            if (bean instanceof AbstractPreAuthenticatedProcessingFilter filter) {
                filter.setMfaEnabled(true);
            }
            if (bean instanceof BasicAuthenticationFilter filter) {
                filter.setMfaEnabled(true);
            }
            return bean;
        }
    };
}
```

(HttpSecurity-built filters go through `AutowireBeanFactoryObjectPostProcessor.initializeBean`, so a `BeanPostProcessor` reaches them.)

Note: `WebAuthnAuthenticationSuccessHandler` is fine as-is — it preserves `authentication.getAuthorities()`, and the merge happens in the filter *before* the success handler runs, so the converted token keeps the merged factors.

## Workaround

Consuming apps can add the missing half themselves (this is what the demo app now does in `MfaSecurityConfig`):

```java
@Configuration
@ConditionalOnProperty(name = "user.mfa.enabled", havingValue = "true")
@EnableMultiFactorAuthentication(authorities = {})
public class MfaSecurityConfig { }
```

The empty `authorities` attribute skips Spring's `AuthorizationManagerFactoryConfiguration` (the framework already provides that bean) and only imports the filter post-processor.

## Related observation

While debugging this, a second integration gap surfaced: `WebSecurityConfig.getUnprotectedURIsList()` auto-unprotects `/user/mfa/status` when MFA is enabled, but **not** the configured `passwordEntryPointUri` / `webauthnEntryPointUri`. If a consuming app forgets to add its WebAuthn entry point page to `unprotectedURIs`, partially-authenticated users are redirected to a page they are denied access to — an infinite redirect loop (`ERR_TOO_MANY_REDIRECTS`). Consider auto-adding the configured entry point URIs the same way `/user/mfa/status` is added. Happy to split this into its own issue if preferred.

## Environment

- ds-spring-user-framework 4.3.1
- Spring Boot 4.0.6 / Spring Security 7.0.4
- Verified end-to-end in SpringUserFrameworkDemoApp PR #60 (fix commit: `MfaSecurityConfig` + Playwright virtual-authenticator E2E that locks the behavior in)

### PR DESCRIPTION
**Pull request by devondragon**
## Summary

Security and reliability hardening for the **4.4.0** release (Spring Boot 4.0 / Spring Security 7). This branch executes the Tier 1 remediation plan, fixes issues found by a follow-up code review, resolves the MFA factor-merging issue (#313), and was validated end-to-end against the demo app (Playwright **103/103**).

**46 commits.** Library suite: **854 tests green**, Javadoc builds, runtime-tested on Java 21 + 25.

### Security fixes
- **RegistrationGuard bypass closed** — internal `persist*` methods are now package-private so consumers can't skip the centralized guard.
- **Verification tokens are single-use** — `validateVerificationToken` now enables the user and deletes the token atomically (was a non-atomic two-step that allowed replay).
- **Facebook OAuth2** now rejects an explicitly-unverified email (parity with the Google path).
- **OAuth2 failure handler** no longer logs raw exception messages (which can embed account emails) at ERROR; expected failures log a non-sensitive summary at WARN.
- **MFA factor merging (#313)** — the framework enables SS7 factor merging on the auth filters itself (no consumer `@EnableMultiFactorAuthentication` workaround needed), and auto-unprotects the configured factor entry-point URIs to prevent the partial-auth redirect loop.

### Reliability / correctness
- **`SecurityFilterChain` coexistence (was a regression)** — the library chain backed off via a *type-based* `@ConditionalOnMissingBean(SecurityFilterChain.class)`, which silently suppressed the entire library chain (form login, CSRF, sessions, WebAuthn, OAuth2) the moment a consumer defined *any* chain — e.g. a narrow actuator/test-API chain. Now keyed on the bean **name** `securityFilterChain`, so additional consumer chains coexist (standard multi-chain `@Order` layering) and a full replacement is an explicit opt-in. *This is what broke ~70 demo Playwright tests until fixed.*
- **Session regeneration on password change** — a self-service password change now preserves and regenerates the **current** session (OWASP) and invalidates only the user's *other* sessions, instead of logging the user out of the device they just used. Toggle: `user.session.invalidation.keep-current-session-on-password-change` (default `true`).
- Security beans (`methodSecurityExpressionHandler`, `httpSessionEventPublisher`, `authenticationEventPublisher`) moved to an auto-config with `@ConditionalOnMissingBean` so consumer overrides no longer conflict.
- MFA filter-merging post-processor isolated into its own dependency-free `@Configuration` to avoid early-instantiation of the event-listening `MfaConfiguration`.

### Infra
- **CI workflow added** (Phase 0): build/test on **Java 21 + 25** plus CodeQL (`security-extended`). Java 17 dropped (SB 4.0 requires 21+).
- **Javadoc jar now builds** (release blocker fixed: registered `@apiNote`/`@implSpec`/`@implNote` tags; fixed a Lombok-getter `@link`).

## Breaking changes (documented in `MIGRATION.md`)
- `UserEmailService` constructor gained a `TokenHasher` parameter (affects subclasses).
- Passay upgraded 1.x → **2.0.0** (package relocations; only affects apps using Passay directly).
- Full-replacement `SecurityFilterChain` must now be named `securityFilterChain`.
- Session-on-password-change behavior change (above), with an opt-out flag.

## Validation
- Library: 854 tests, 0 failures; Javadoc builds; `testJdk25` green.
- Demo app (consuming the local snapshot): boots clean, **Playwright 103/103**, MFA integration tests green with the consumer workaround removed.
- The `SecurityFilterChain` regression and the session/MFA fixes were each caught or confirmed by running the demo app end-to-end.

Closes #313.

### CODE REVIEW

File: src/main/java/com/digitalsanctuary/spring/user/service/UserEmailService.java [hunk 1]
```diff
@@ -200,11 +208,21 @@ private String generateToken() {
...
      */
+    @Transactional
     public void createPasswordResetTokenForUser(final User user, final String token) {
-        final PasswordResetToken myToken = new PasswordResetToken(token, user);
+        // Single active token per user: remove any previously issued token before creating a new one.
+        passwordTokenRepository.deleteByUser(user);
+        // Store only the hash of the token; the raw token is what was emailed to the user.
+        final PasswordResetToken myToken =
+                new PasswordResetToken(tokenHasher.hash(token), user, passwordResetTokenValidityMinutes);
         passwordTokenRepository.save(myToken);
```
[the remark this answers is not in the document]
Comment (devondragon) [reply src/main/java/com/digitalsanctuary/spring/user/service/UserEmailService.java:246]: Good catch — confirmed the `@Transactional` was a no-op via self-invocation. Fixed in the latest commit: added a `@Lazy @Autowired UserEmailService self` proxy reference (mirroring the existing pattern in `UserService`) and routed both callers (`sendForgotPasswordVerificationEmail` and `initiateAdminPasswordReset`) through `self.createPasswordResetTokenForUser(...)` so the `deleteByUser` + `save` now commit atomically. Updated `UserEmailServiceTest` to wire `self` to the instance (no proxy in unit tests).

File: src/main/java/com/digitalsanctuary/spring/user/service/LoginHelperService.java [hunk 2]
```diff
@@ -96,7 +101,31 @@ public DSUserDetails userLoginHelper(User dbUser, OidcUserInfo oidcUserInfo, Oid
...
+    private void assertAccountUsable(User user) {
+        if (user.isLocked()) {
+            log.debug("Rejecting authentication for locked account: {}", user.getEmail());
+            throw new LockedException("Account is locked: " + user.getEmail());
+        }
+        if (!user.isEnabled()) {
+            log.debug("Rejecting authentication for disabled account: {}", user.getEmail());
+            throw new DisabledException("Account is disabled: " + user.getEmail());
+        }
+    }
```
[the remark this answers is not in the document]
Comment (devondragon) [reply src/main/java/com/digitalsanctuary/spring/user/service/LoginHelperService.java:132]: Agreed — fixed. The `LockedException`/`DisabledException` messages are now generic (`"Account is locked"` / `"Account is disabled"`) with no PII, and the user's email is retained only in the existing DEBUG log lines. This removes the dependency on every downstream handler sanitizing the message.

**Comment by devondragon** [comment]
## Review feedback addressed

Thanks for the thorough review. Pushed fixes for the actionable items:

| # | Item | Fix |
|---|------|-----|
| 1 | **PII in `LockedException`/`DisabledException`** | Messages are now generic (`"Account is locked"` / `"Account is disabled"`); email retained only in DEBUG logs. Removes the dependency on downstream handlers sanitizing the message. |
| 2 | **Null guard return in `CompositeRegistrationGuard.evaluate`** | Now throws `IllegalStateException` on a null delegate decision instead of failing open. Added two unit tests covering the fail-fast behavior. |
| 3 | **`FileAuditLogFlushScheduler` SpEL default mismatch** | Changed `!${user.audit.flushOnWrite:true}` → `:false` to match `AuditConfig`'s Java field default, so the scheduler is correctly created when the property is absent. |
| 4 | **`tokenFingerprint` duplication** | Extracted to `public static TokenHasher.fingerprint(...)`; removed the identical private copies in `UserVerificationService` and `UserActionController`. |
| 5 | **`cleanUpPasswordHistory` comment** | Rewritten to reflect the real call chain (`changeUserPassword` → `self.persistChangedPassword` (proxied) → `savePasswordHistory` → `cleanUpPasswordHistory`); removed the misleading "bypassed by proxy" claim. |

Plus the must-fix Copilot caught: the `@Transactional` on `createPasswordResetTokenForUser` was a no-op via self-invocation — now routed through a `@Lazy self` proxy so the delete+save commit atomically.

**Deferred / declined:**
- **Guard-before-duplicate-check bcrypt amplification** — valid, but moving the fast-path duplicate check ahead of the guard is a design change out of scope for this PR. Tracking separately; expected rate limiting at the controller/gateway mitigates in the interim.
- **`email_verified` unknown-type debug log** and **`UserService` subclass double-guard edge case** — declined as low-signal / theoretical API-misuse; behavior is intentional and documented inline.

All affected test suites pass.

🤖 Generated with [Claude Code](https://claude.ai/claude-code)

**Comment by devondragon** [comment]
## Second review pass addressed (commit `5a7578b`)

Thanks @claude — caught real issues on the re-review. Triaged and fixed the valid ones:

### 🔴 Fixed before merge
| Item | Fix |
|------|-----|
| **#1 Session created on every OAuth2 failure** | `SanitizingOAuth2AuthenticationFailureHandler` now uses `request.getSession(false)` + null-guard, so a sessionless caller (unauthenticated scanner hitting the callback) no longer forces session allocation. A genuine OAuth2 flow already has a session from the authorization-request phase, so the user-facing message is preserved. Tests updated to model the real path; added a test asserting no session is created for a cold request. |
| **#2 NPE in `generateNewVerificationToken`** | Added a null guard on the resolved token; now throws `IllegalArgumentException` with a clear message instead of NPEing on `updateToken(...)`. |
| **CodeQL log-injection ×3** (advanced-security bot) | `TokenHasher.fingerprint()` now strips control chars (CR/LF) from the logged prefix, closing the log-forging vector centrally for all callers. |

### 🟡 Quick wins
- **#5** Converted tab indentation → 4 spaces in `MfaFilterMergingConfiguration` (CLAUDE.md style).

### Already covered / declining
- **#6 (NOT_SUPPORTED migration note)** — already documented in `MIGRATION.md` § *Password hashing moved outside the transaction*, including the bold "an outer rollback will not roll back the registration or password change" caveat. No change needed.
- **#3 SHA-pin CI actions** — valid hardening, but deferring to a follow-up: the actions were just bumped this PR, and pinning every third-party action to a SHA (with a maintenance story) is its own task.
- **#4 startup `log.warn` for MFA merging** — low value vs. the existing class-level JavaDoc + per-filter DEBUG log; skipping.

Full test suite green locally.

🤖 Generated with [Claude Code](https://claude.ai/claude-code)
